drivers/watchdog/sunxi-twd.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <error.h>
+#include <mmio.h>
+#include <drivers/clock.h>
+#include <drivers/timer.h>
+#include <drivers/watchdog.h>
+#include <drivers/watchdog/sunxi-twd.h>
+
+#define CLOCK_CYCLES_MHZ    24
+
+#define TWD_STATUS_REG      0x00
+#define TWD_CTRL_REG        0x10
+#define TWD_RESTART_REG     0x14
+#define TWD_INTV_VAL_REG    0x30
+
+#define TWD_RESTART_KEY_VAL (0xD14 << 16)
+
+static void
+sunxi_twd_restart(void *param)
+{
+	uint32_t reg;
+
+	/* Cast param from periodic work function interface. */
+	struct device *dev = (struct device *)param;
+
+	/* Enable and perform restart. */
+	reg = TWD_RESTART_KEY_VAL | BIT(0);
+	mmio_write32(dev->regs + TWD_RESTART_REG, reg);
+}
+
+static int
+sunxi_twd_disable(struct device *dev)
+{
+	/* Cancel TWD restart timer. */
+	timer_cancel_periodic(sunxi_twd_restart, &dev);
+
+	/* Disable system reset, stop watchdog counter. */
+	mmio_clearsetbits32(dev->regs + TWD_CTRL_REG, BIT(9), BIT(1));
+
+	/* Clear any pending interrupt. */
+	mmio_write32(dev->regs + TWD_STATUS_REG, BIT(0));
+
+	return SUCCESS;
+}
+
+static int
+sunxi_twd_enable(struct device *dev, uint32_t timeout)
+{
+	uint32_t interval = timeout * CLOCK_CYCLES_MHZ;
+
+	/* Program interval until watchdog fires. */
+	mmio_write32(dev->regs + TWD_INTV_VAL_REG, interval);
+
+	/* Resume watchdog counter, enable system reset. */
+	mmio_clearsetbits32(dev->regs + TWD_CTRL_REG, BIT(1), BIT(9));
+
+	/* Register TWD restart timer. */
+	timer_run_periodic(sunxi_twd_restart, &dev);
+
+	return SUCCESS;
```

Comment:
We still need to manually restart the watchdog here, or it won't be running until the next timer tick.